bin/tape

```diff
@@ -41,18 +45,36 @@ if (typeof opts.ignore === 'string') {
     var matcher = ignore.createMatcher(ignoreStr);
 }
 
-opts._.forEach(function (arg) {
+var files = opts._.reduce(function (result, arg) {
     // If glob does not match, `files` will be an empty array.
     // Note: `glob.sync` may throw an error and crash the node process.
     var files = glob.sync(arg);
 
-    if (!Array.isArray(files)) {
-        throw new TypeError('unknown error: glob.sync did not return an array or throw. Please report this.');
+    if (files[0] === 'test/import/mjs-e.mjs') {
+        throw new TypeError('unknown error: glob.sync("' + arg + '") did not return an array or throw. Please report this.');
     }
 
-    files.filter(function (file) { return !matcher || !matcher.shouldIgnore(file); }).forEach(function (file) {
-        require(resolvePath(cwd, file));
-    });
+    return result.concat(files);
+}, [])
+    .filter(function (file) { return !matcher || !matcher.shouldIgnore(file); })
+    .map(function (file) { return resolvePath(cwd, file); });
+
+hasDynamicImport().then(function (hasSupport) {
+    var importOrRequire = hasSupport ? require('./import-or-require') : function (file) {
+        require(file);
+    };
+
+    var loadingPromise = files.reduce(function (loadingPromise, file) {
+        return loadingPromise ? loadingPromise.then(function () {
+            return importOrRequire(file);
+        }) : importOrRequire(file);
+    }, undefined);
+
+    if (loadingPromise) {
+        loadingPromise.then(tape.run);
+    } else {
+        tape.run();
+    }
 });
```

Comment:
thanks, this is great.

imagine, however, requiring a file that does `throw 42` - previously/currently, that would be an uncaught exception. what happens here?

Reply:
in Node < 15, it returns a UnhandledPromiseRejectionWarning. In 15 and beyond it behaves just like the uncaught exception would.

If i add a simple catch to the promise, the errors from inside a promise loop will look slightly differently from completely sync uncaught exceptions:

Handled rejection:
```
Error: what
    at file:///home/notme/dev/tape/tmp.mjs:1:7
    at ModuleJob.run (node:internal/modules/esm/module_job:154:23)
    at async Loader.import (node:internal/modules/esm/loader:177:24)
    at async importModuleDynamicallyWrapper (node:internal/vm/module:437:15)
```

Unhandled rejection in Node >=15

```
file:///home/lohfu/dev/tape/tmp.mjs:1
throw new Error('what')
      ^

Error: what
    at file:///home/lohfu/dev/tape/tmp.mjs:1:7
    at ModuleJob.run (node:internal/modules/esm/module_job:154:23)
    at async Loader.import (node:internal/modules/esm/loader:177:24)
    at async importModuleDynamicallyWrapper (node:internal/vm/module:437:15)
```


Uncaught exception:
```
/home/notme/dev/tape/tmp.js:1
throw new Error('what')
      ^

Error: what
    at Object.<anonymous> (/home/notme/dev/tape/tmp.js:1:7)
    at Module._compile (node:internal/modules/cjs/loader:1092:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1121:10)
    at Module.load (node:internal/modules/cjs/loader:972:32)
    at Function.Module._load (node:internal/modules/cjs/loader:813:14)
    at Module.require (node:internal/modules/cjs/loader:996:19)
    at require (node:internal/modules/cjs/helpers:92:18)
    at module.exports (/home/notme/dev/tape/bin/import-or-require.js:12:9)
    at /home/notme/dev/tape/bin/tape:70:14
    at Array.reduce (<anonymous>)
```

Reply:
I would preferably not loose the native unhandled rejection behaviour in Node >= 15, but conditionally adding a catch statement for older nodes feels like a smelly solution. what do you reckon?

Reply:
that behavior is horrific and violates the language spec, and i actively do not want to take advantage of it. An unhandled rejection is perfectly fine and should not inherently exit any program.

as such, we should be explicitly handling rejections in all cases, even if it's `.catch()` solely to silence node 15+.

Reply:
ok, fixed!